ocaml/xapi/vgpuops.ml

```diff
@@ -125,7 +125,17 @@ let reserve_free_virtual_function ~__context vm impl pf =
         (* This probably means that our capacity checking went wrong! *)
         raise Api_errors.(Server_error (internal_error, ["No free virtual function found"]))
   in
-  get true
+  match impl with
+  | `nvidia_sriov ->
+    (* Always call the setup function. If we are coming out of
+     * pass-through, xapi assumes the VFs are active but they are not - so
+     * activate them. The setup function itself avoids unnecessary
+     * work *)
+    Helpers.call_api_functions ~__context @@ fun rpc session_id ->
+        Client.Client.Host.nvidia_vf_setup ~rpc ~session_id ~host ~pf
```

Comment:
I thought that it would be better to call the Xapi_host method directly here, but you might need some additional locking (see below), so this is probably better as it is written currently.

Reply:
@edwintorok, the functions in this module always run on the master, while the nvidia setup needs to be done on the host that will run the VM. So we have to go through the API forwarding in any case.